Remotion/Web/Core/Resources/ResourcePathBuilderBasedStaticResourceCacheKeyProvider.cs

```diff
@@ -0,0 +1,251 @@
+          foreach (var resourceType in resourceTypes)
+          {
+            var resourceTypeDirectory = new DirectoryInfo(Path.Combine(assemblyFolder, resourceType.Name));
+            if (!resourceTypeDirectory.Exists)
+              continue;
+
+            foreach (var resourceFile in resourceTypeDirectory.EnumerateFiles("*", SearchOption.AllDirectories))
+            {
+              stringBuilder.Append(resourceFile.FullName);
+              stringBuilder.Append(';');
+              stringBuilder.Append(resourceFile.Length);
+              stringBuilder.Append(';');
+              stringBuilder.Append(resourceFile.CreationTimeUtc.ToFileTimeUtc());
+              stringBuilder.Append(';');
+              stringBuilder.Append(resourceFile.LastWriteTimeUtc.ToFileTimeUtc());
+              stringBuilder.AppendLine();
+            }
+          }
+        }
+      }
+    }
+
+    private enum FileChangeType
+    {
+      Changed,
+      Created,
+      Deleted,
+      Renamed
+    }
+
+    private readonly IResourcesPhysicalPathLocator _resourcesPhysicalPathLocator;
+    private readonly IResourceFileDetailsAppender _resourceFileDetailsAppender;
+    private readonly IReadOnlyList<ResourceType> _resourceTypes;
+
+    private readonly HashSet<string> _resourceTypeLookup;
+
+    private readonly object _lock = new();
+
+    private string? _physicalPath;
+    private FileSystemWatcher? _fileSystemWatcher;
+
+    // Stores the current cacheKey or null if no cache key has been generated
+    // Set to null without the lock, set to not-null always in the lock.
+    private string? _cacheKey;
+
+    public ResourcePathBuilderBasedStaticResourceCacheKeyProvider (IHttpContextProvider httpContextProvider, ResourceRoot resourceRoot)
+        : this(
+            new ResourcesPhysicalPathLocator(
+                ArgumentUtility.CheckNotNull("httpContextProvider", httpContextProvider),
+                ArgumentUtility.CheckNotNull("resourceRoot", resourceRoot)),
+            new ResourceFileDetailsAppender(),
+            ResourceType.CacheableResourceTypes)
+    {
+    }
+
+    protected ResourcePathBuilderBasedStaticResourceCacheKeyProvider (
+        IResourcesPhysicalPathLocator resourcesPhysicalPathLocator,
+        IResourceFileDetailsAppender resourceFileDetailsAppender,
+        IReadOnlyList<ResourceType> resourceTypes)
+    {
+      ArgumentUtility.CheckNotNull(nameof(resourcesPhysicalPathLocator), resourcesPhysicalPathLocator);
+      ArgumentUtility.CheckNotNull(nameof(resourceFileDetailsAppender), resourceFileDetailsAppender);
+      ArgumentUtility.CheckNotNull(nameof(resourceTypes), resourceTypes);
+
+      _resourcesPhysicalPathLocator = resourcesPhysicalPathLocator;
+      _resourceFileDetailsAppender = resourceFileDetailsAppender;
+      _resourceTypes = resourceTypes;
+
+      _resourceTypeLookup = resourceTypes.Select(e => e.Name).ToHashSet(StringComparer.OrdinalIgnoreCase);
+    }
+
+    public bool HasCacheKey => _cacheKey != null;
```

Comment:
We already have the DoubleCheckedLockingContainer for this. While I want to eventually phase it out and replace it with lower level APIs *), until such a time, we can use it when having a writable synchronization istuation.

it works the same. you set the implementation delegate in the ctor and when the value is reset to null, it will get reevaluated.

*) most scenarios where we use DCLC are write once or should be refactored to write once and can thus be covered by Lazy<T> these days. The couple of read/write places that remain can then be written with manual lock management.